fabric-events-lifecycle-v0/src/main/java/net/fabricmc/fabric/mixin/eventslifecycle/MixinWorldRenderer.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.mixin.eventslifecycle;
+
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+import net.fabricmc.fabric.api.event.client.InvalidateRenderStateCallback;
+import net.minecraft.client.render.WorldRenderer;
+
+@Mixin(WorldRenderer.class)
+public abstract class MixinWorldRenderer {
+    @Inject(method = "reload", at = @At("HEAD"))
```

Comment:
Mojang only reloads when the world is non-null. This hook runs regardless. The injection point needs to be more precise to support this.

Reply:
I sincerely appreciate the thoughtful analysis. I fear my first paragraph is leading you astray.  Changing the event to depend on the presence or absence of World or any other state would be inconsistent with its intended purpose. 

The reload() method accepts no parameters and the second paragraph of the event doc states: 
```
 Render chunks and other render-related object instances will be made null
 or invalid after this event so do not use it to capture dependent state.
 Instead, use it to invalidate state and reinitialize lazily.
```

The event makes two useful guarantees:

1) Render state during and after this event is or may become invalid and any state currently derived from it must be invalidated.  It offers _no_ assurances about when, or even _if_, valid render state will again become available.

2) It happens early enough for any necessary invalidation or reinitialization to happen before the render loop restarts, if it ever does.  It could also be said to happen _outside_ the render loop.

The only state that can be reliably accessed during this event is state that is internal to the mod itself and independent of game state. (Configuration, mostly.). Earlier in the thread I confirmed all my current use cases are consistent with this pattern: clearing collections, invalidating caches, reseting internal counters, re-deriving static configuration components, etc.

Certainly, any attempt to access world state at this point, even if a non-null world object is present, should be discouraged - there is nothing to prevent render-related state in the world instance from being mutated after this event is called.  The reliable way to know that render state is valid is by reacting to the normal sequence of method calls that occur during the render loop, at the times they are called.

The location of the hook in WorldRenderer.reload() is not part of the contract - that merely happens to be a convenient place to implement it currently. There is no guarantee that will remain true as the game changes.  However, the contract of this event as described should remain durable and relevant wherever the hook lives.  

Given all that, I think I should reduce or change the first paragraph to avoid any specific reference to WorldRenderer or other implementation details.

Reply:
As it stands, this is not called on invalidation. It is called when the renderer is *rebuilt*. If you wish to handle all forms of invalidation with the API user handling world nullity, you should also be invoking this callback from `WorldRenderer#setWorld` in the case of a null world where Mojang invalidates the world renderer data.